fuse_models/test/test_unicycle_2d_state_cost_function.cpp

```diff
@@ -0,0 +1,141 @@
+ *  are met:
+ *
+ *   * Redistributions of source code must retain the above copyright
+ *     notice, this list of conditions and the following disclaimer.
+ *   * Redistributions in binary form must reproduce the above
+ *     copyright notice, this list of conditions and the following
+ *     disclaimer in the documentation and/or other materials provided
+ *     with the distribution.
+ *   * Neither the name of the copyright holder nor the names of its
+ *     contributors may be used to endorse or promote products derived
+ *     from this software without specific prior written permission.
+ *
+ *  THIS SOFTWARE IS PROVIDED BY THE COPYRIGHT HOLDERS AND CONTRIBUTORS
+ *  "AS IS" AND ANY EXPRESS OR IMPLIED WARRANTIES, INCLUDING, BUT NOT
+ *  LIMITED TO, THE IMPLIED WARRANTIES OF MERCHANTABILITY AND FITNESS
+ *  FOR A PARTICULAR PURPOSE ARE DISCLAIMED. IN NO EVENT SHALL THE
+ *  COPYRIGHT HOLDER OR CONTRIBUTORS BE LIABLE FOR ANY DIRECT, INDIRECT,
+ *  INCIDENTAL, SPECIAL, EXEMPLARY, OR CONSEQUENTIAL DAMAGES (INCLUDING,
+ *  BUT NOT LIMITED TO, PROCUREMENT OF SUBSTITUTE GOODS OR SERVICES;
+ *  LOSS OF USE, DATA, OR PROFITS; OR BUSINESS INTERRUPTION) HOWEVER
+ *  CAUSED AND ON ANY THEORY OF LIABILITY, WHETHER IN CONTRACT, STRICT
+ *  LIABILITY, OR TORT (INCLUDING NEGLIGENCE OR OTHERWISE) ARISING IN
+ *  ANY WAY OUT OF THE USE OF THIS SOFTWARE, EVEN IF ADVISED OF THE
+ *  POSSIBILITY OF SUCH DAMAGE.
+ */
+#include <fuse_models/unicycle_2d_state_cost_function.h>
+#include <fuse_models/unicycle_2d_state_cost_functor.h>
+
+#include <gtest/gtest.h>
+#include <fuse_core/eigen_gtest.h>
+
+#include <ceres/autodiff_cost_function.h>
+#include <ceres/gradient_checker.h>
+#include <Eigen/Dense>
+
+#include <limits>
+#include <vector>
+
+
+TEST(CostFunction, evaluateCostFunction)
+{
+  // Create cost function
+  const double process_noise_diagonal[] = { 1e-3, 1e-3, 1e-2, 1e-6, 1e-6, 1e-4, 1e-9, 1e-9 };
+  const fuse_core::Matrix8d covariance = fuse_core::Vector8d(process_noise_diagonal).asDiagonal();
+
+  const double dt{ 0.1 };
+  const fuse_core::Matrix8d sqrt_information{ covariance.inverse().llt().matrixU() };
+
+  const fuse_models::Unicycle2DStateCostFunction cost_function{ dt, sqrt_information };
+
+  // Evaluate cost function
+  const double position1[] = {0.0, 0.0};
+  const double yaw1[] = {0.0};
+  const double vel_linear1[] = {1.0, 0.0};
+  const double vel_yaw1[] = {1.570796327};
+  const double acc_linear1[] = {1.0, 0.0};
+
+  const double position2[] = {0.105, 0.0};
+  const double yaw2[] = {0.1570796327};
+  const double vel_linear2[] = {1.1, 0.0};
+  const double vel_yaw2[] = {1.570796327};
+  const double acc_linear2[] = {1.0, 0.0};
+
+  const double* parameters[] =
+  {
+    position1, yaw1, vel_linear1, vel_yaw1, acc_linear1,
+    position2, yaw2, vel_linear2, vel_yaw2, acc_linear2
+  };
+
+  fuse_core::Vector8d residuals;
+
+  const auto& block_sizes = cost_function.parameter_block_sizes();
```

Comment:
Again, here we convert from `ceres` jacobian layout to a full jacobian. But in this case we take the parameter block sizes from the cost function. This is NOT available at compile time, so we must use an `std::vector` instead of an `std::array`. If we create a conversion util, it should ideally support both the compile (array) and runtime (vector) cases. I believe this is an important point that should be considered. Maybe this could also be done at compile time if we access the parameter block sizes from the class, rather than an instantiated object. :thinking: